src/journal/journal-file.c

```diff
@@ -1421,14 +1469,28 @@ int journal_file_find_data_object_with_hash(
                         if (ret)
                                 *ret = o;
 
-                        if (offset)
-                                *offset = p;
+                        if (ret_offset)
+                                *ret_offset = p;
 
                         return 1;
                 }
 
         next:
-                p = le64toh(o->data.next_hash_offset);
+                nextp = le64toh(READ_NOW(o->data.next_hash_offset));
+                if (nextp > 0) {
+
+                        if (nextp <= p) /* Refuse going in loops */
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                       "Detected data hash item loop in %s, refusing.", f->path);
+
+                        depth++;
+
+                        /* If the depth of this hash chain is larger than all others we have seen so far, record it */
+                        if (JOURNAL_HEADER_CONTAINS(f->header, data_hash_chain_depth) && f->writable)
+                                f->header->data_hash_chain_depth = htole64(MAX(depth, le64toh(f->header->data_hash_chain_depth)));
```

Comment:
Like with the field_hash_chain_depth, this will issue writes on every iteration rather than only when the value needs to be changed.

Reply:
again, this is up to the compiler and OS to optimize. Doing this in a local variable first and comparing before copying it over should not make a difference.